### PR TITLE
tests: drivers: dac: add dac testcode

### PR DESCRIPTION
**Pull request by Saba-AlifSemi**
tests: drivers: dac: add initial DAC QA test suite for sdk-alif

What’s included
New DAC functional test coverage for driver behavior and stability.
Capacitance-focused test scenarios, including ramp behavior with settling checks.
Validation of DAC output current configuration handling.
Runtime assertions and logs to help QA quickly identify failures and root causes.

### CODE REVIEW

**Review by MaheshYerra-AlifSemi — COMMENTED** [review]
Looks okay for me

File: tests/drivers/dac/src/main.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/main.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac.h]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_basic.c [hunk 3]
```diff
@@ -0,0 +1,591 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_boundary.c [hunk 4]
```diff
@@ -0,0 +1,378 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_boundary.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_capacitance.c [hunk 5]
```diff
@@ -0,0 +1,342 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_capacitance.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_data_format.c [hunk 6]
```diff
@@ -0,0 +1,385 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_data_format.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_continuous.c [hunk 7]
```diff
@@ -0,0 +1,418 @@
+/* Copyright (C) 2026 Alif Semiconductor - All Rights Reserved.
+ * Use, distribution and modification of this code is permitted under the
+ * terms stated in the Alif Semiconductor Software License Agreement
+ *
+ * You should have received a copy of the Alif Semiconductor Software
+ * License Agreement with this file. If not, please write to:
+ * contact@alifsemi.com, or visit: https://alifsemi.com/license
+ *
+ * SPDX-License-Identifier: LicenseRef-Alif-Proprietary
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_continuous.c]: License Header should not have SPDX-License

File: tests/drivers/dac/src/test_dac_basic.c [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+		.resolution = DAC_RESOLUTION,
+		.buffered = 0
+	};
+
+	ret = dac_channel_setup(dev, &cfg);
+	zassert_equal(ret, 0, "DAC channel setup failed [%d]", ret);
+
+	LOG_INF("DAC channel setup test PASSED");
+}
+
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c]: I think this same functionality is done in init_dac_device(). Please merge the functionality into one; do not duplicate

File: tests/drivers/dac/src/test_dac_basic.c [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+/* -------- Test: Dual DAC independent write (R-32_1) -------- */
+static void dac_dual_channel_independent_write_test(void)
+{
+	int ret;
+
+	LOG_INF("=== DAC Dual Channel Independent Write Test ===");
+
+#if DT_NODE_HAS_STATUS(DAC0_NODE, okay) && DT_NODE_HAS_STATUS(DAC1_NODE, okay)
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c:177]: This does not belong here and is simply overhead of checking. Move it to the init sequence

File: tests/drivers/dac/src/test_dac_basic.c [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+}
+
+/* -------- Test: Dual DAC simultaneous operation (R-32_1) -------- */
+static void dac_dual_channel_simultaneous_test(void)
+{
+	int ret;
+
+	LOG_INF("=== DAC Dual Channel Simultaneous Test ===");
+
+#if DT_NODE_HAS_STATUS(DAC0_NODE, okay) && DT_NODE_HAS_STATUS(DAC1_NODE, okay)
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c:219]: This does not belong here and is simply overhead of checking. Move it to the init sequence

File: tests/drivers/dac/src/main.c [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+const struct device *dac_dev;
+
+/* Default channel configuration: 12-bit, channel 0, unbuffered */
+const struct dac_channel_cfg dac_ch_cfg = {
+	.channel_id = DAC_CHANNEL_ID,
+	.resolution = DAC_RESOLUTION,
+	.buffered = 0
+};
+
+const struct device *init_dac_device(void)
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/main.c]: Can this function be called in setup, as it returns a pointer?

**Review by sudiiyengar-en — CHANGES_REQUESTED** [review]

File: tests/drivers/dac/testcase.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+      - alif_e7_dk/ae722f80f55d5xx/rtss_hp
+    harness: ztest
+    timeout: 600
+    integration_platforms:
+      - alif_b1_dk/ab1c1f4m51820ph0/rtss_he
+      - alif_e1c_dk/ae1c1f4051920hh/rtss_he
+      - alif_e8_dk/ae822fa0e5597xx0/rtss_he
+      - alif_e8_dk/ae822fa0e5597xx0/rtss_hp
+      - alif_e7_dk/ae722f80f55d5xx/rtss_he
+      - alif_e7_dk/ae722f80f55d5xx/rtss_hp
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/testcase.yaml]: Please run checkpatch

**Comment by sudiiyengar-en** [comment]
The boundary test cases are almost similar, so my suggestion is to reduce the cases and keep the scenario in one, so one test case should cover all possibilities

File: tests/drivers/dac/src/main.c [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+	int ret;
+	const struct device *dev = DEVICE_DT_GET(DAC_NODE);
+
+	zassert_not_null(dev, "DAC device handle is NULL");
+	zassert_true(device_is_ready(dev), "DAC device is not ready");
+
+	ret = dac_channel_setup(dev, &dac_ch_cfg);
+	zassert_equal(ret, 0, "DAC channel setup failed [%d]", ret);
+
+	return dev;
```
Comment (Nagendra-AlifSemi) [inline tests/drivers/dac/src/main.c]: use return type to void * (cast the const struct device *) si it matches ztest's setup signature and retrieve it via the fixture.
fix: return (void *)dev;

File: tests/drivers/dac/src/test_dac_basic.c [hunk 14]
```diff
@@ -0,0 +1,591 @@
...
+ *   - 12-bit resolution
+ *   - Internal 1.8V voltage reference
+ */
+
+#include "test_dac.h"
+
+LOG_MODULE_DECLARE(dac_test, LOG_LEVEL_INF);
+
+/* -------- Test: DAC device is ready -------- */
+static void dac_device_ready_test(void)
```
Comment (Nagendra-AlifSemi) [inline tests/drivers/dac/src/test_dac_basic.c]: Remove the redundant tests entirely
Since test_dac_basic_setup already calls device-ready checks (and once you adopt init_dac_device() there, channel setup is also covered), these standalone tests add no coverage. Delete dac_device_ready_test, dac_channel_setup_test, and their ZTEST wrappers.

else 
Keep the tests but route them through init_dac_device()
static void dac_device_ready_and_setup_test(void)
{
    LOG_INF("=== DAC Device Ready + Channel Setup Test ===");
    dac_dev = init_dac_device();   /* asserts ready + channel setup OK */
    zassert_not_null(dac_dev, "init_dac_device() returned NULL");
    LOG_INF("DAC device ready + setup test PASSED");
}

ZTEST(test_dac_basic, test_dac_device_ready_and_setup)
{
    dac_device_ready_and_setup_test();
}

File: tests/drivers/dac/src/test_dac_basic.c [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+
+/* -------- Test: Dual DAC independent write (R-32_1) -------- */
+static void dac_dual_channel_independent_write_test(void)
+{
+	int ret;
+
+	LOG_INF("=== DAC Dual Channel Independent Write Test ===");
+
+#if DT_NODE_HAS_STATUS(DAC0_NODE, okay) && DT_NODE_HAS_STATUS(DAC1_NODE, okay)
+	const struct device *dac0 = DEVICE_DT_GET(DAC0_NODE);
```
Comment (Nagendra-AlifSemi) [inline tests/drivers/dac/src/test_dac_basic.c]: introduce a dual-init helper (mirror of init_dac_device()), keep it in the same place that init_dac_device() lives
**/* in test_dac.h */**
struct dac_dual_devs {
    const struct device *dac0;
    const struct device *dac1;
};

#if DT_NODE_HAS_STATUS(DAC0_NODE, okay) && DT_NODE_HAS_STATUS(DAC1_NODE, okay)
struct dac_dual_devs init_dac_dual_devices(void);
#endif

**/* in test_dac.c */**
#if DT_NODE_HAS_STATUS(DAC0_NODE, okay) && DT_NODE_HAS_STATUS(DAC1_NODE, okay)
struct dac_dual_devs init_dac_dual_devices(void)
{
    int ret;
    struct dac_dual_devs d = {
        .dac0 = DEVICE_DT_GET(DAC0_NODE),
        .dac1 = DEVICE_DT_GET(DAC1_NODE),
    };
    const struct dac_channel_cfg cfg = {
        .channel_id = DAC_CHANNEL_ID,
        .resolution = DAC_RESOLUTION,
        .buffered   = 0,
    };

    zassert_not_null(d.dac0, "DAC0 device handle is NULL");
    zassert_not_null(d.dac1, "DAC1 device handle is NULL");
    zassert_true(device_is_ready(d.dac0), "DAC0 not ready");
    zassert_true(device_is_ready(d.dac1), "DAC1 not ready");

    ret = dac_channel_setup(d.dac0, &cfg);
    zassert_equal(ret, 0, "DAC0 channel setup failed [%d]", ret);
    ret = dac_channel_setup(d.dac1, &cfg);
    zassert_equal(ret, 0, "DAC1 channel setup failed [%d]", ret);

    return d;
}
#endif

**Review by Nagendra-AlifSemi — COMMENTED** [review]
changes required

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
The overlay files belong at the samples root-level dir - not on the boards directory. The boards-directory is meant for board specific overlay files

File: tests/drivers/dac/src/test_dac_basic.c [hunk 16]
```diff
@@ -0,0 +1,517 @@
...
+
+	ret = dac_channel_setup(dev, &cfg_ch1);
+	zassert_equal(ret, -EINVAL,
+			  "Expected -EINVAL for channel_id=1, got [%d]", ret);
+
+	/* Test channel_id = 255 (invalid - max uint8) */
+	struct dac_channel_cfg cfg_ch255 = {
+		.channel_id = 255,
+		.resolution = DAC_RESOLUTION,
+		.buffered = 0
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c:273]: dac_twoscomp_input_range_test asserts conflicting behavior
The test claims to validate over-range rejection in two's-complement mode but then asserts ret == 0 for both 0x1FFF and 0xFFFF. This locks in the buggy behavior described in R-BUG-2 as "passing", so the test will silently pass forever and won't fail when the driver is fixed.

Suggest using ztest_test_skip() or TC_PRINT with a known-failure marker, or invert the assertion to fail until R-BUG-2 is fixed. As written, this is a regression hazard.

File: tests/drivers/dac/src/test_dac_basic.c [hunk 17]
```diff
@@ -0,0 +1,517 @@
...
+	/* Test value exceeding 12-bit range in twoscomp mode */
+	/* Driver should either mask to 12 bits or return -EINVAL */
+	ret_1 = dac_write_value(dac_dev, DAC_CHANNEL_ID, 0x1FFF);
+	/* Currently driver accepts any value silently; this test documents behavior */
+	/* After fix: zassert_equal(ret, -EINVAL, ...) or verify value is masked */
+
+	/* Test 16-bit over-range value */
+	ret_2 = dac_write_value(dac_dev, DAC_CHANNEL_ID, 0xFFFF);
+	/* Same - documents current behavior vs expected */
+
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c:301]: dac_rapid_write_stress_test is not actually a stress test
With k_sleep(K_MSEC(50)) between every write, 200 writes take ~10 seconds and there is no high-frequency stress on the driver. Either remove the delays (the comment says "for LA visibility" but this is a CI/QA test, not a bench demo), or rename the test to reflect what it actually does. The same issue applies to dac_continuous_rapid_writes_test in test_dac_continuous.c:222-243.

File: tests/drivers/dac/src/test_dac_basic.c [hunk 18]
```diff
@@ -0,0 +1,517 @@
...
+}
+
+/* -------- Test: Basic DAC write value -------- */
+static void dac_basic_write_test(void)
+{
+	int ret;
+
+	LOG_INF("=== DAC Basic Write Test ===");
+
+	dac_dev = init_dac_device();
```
Comment (sudiiyengar-en) [inline tests/drivers/dac/src/test_dac_basic.c]: init_dac_device() in every test
Every test function calls dac_dev = init_dac_device(); at the top, but test_dac_basic_setup already does this once. The per-test re-init is redundant and re-runs dac_channel_setup unnecessarily. Either remove from individual tests or move to before if a fresh state is required between tests.

File: tests/drivers/dac/testcase.yaml
Comment (sudiiyengar-en) [inline tests/drivers/dac/testcase.yaml]: The testcase.yaml declares no extra_args for DTC_OVERLAY_FILE, so twister will build with no overlay applied, and DT_ALIAS(dac) will likely not resolve (no dac alias defined in base DT). The suite as configured won't produce a working binary in CI. Add per-platform overlay variants

File: tests/drivers/dac/src/test_dac_boundary.c [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+
+	LOG_INF("=== DAC Boundary Min Non-Zero (0x001) Test ===");
+
+	dac_dev = init_dac_device();
+
+	ret = dac_write_value(dac_dev, DAC_CHANNEL_ID, 0x001);
+	zassert_equal(ret, 0, "DAC write 0x001 failed [%d]", ret);
+	k_sleep(K_MSEC(100));
+
+	LOG_INF("DAC boundary min non-zero test PASSED");
```
Comment (VeijoPesonen) [inline tests/drivers/dac/src/test_dac_boundary.c]: Should TC_PRINT be used instead?

File: tests/drivers/dac/README.rst [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    west build -b alif_e8_dk/ae822fa0e5597xx0/rtss_he ../alif/tests/drivers/dac/ \
+    -- -DDTC_OVERLAY_FILE=alif_dac0_twoscomp.overlay
+
+    # Maximum output current (1.5 mA) - HRM feature test
+    west build -b alif_e8_dk/ae822fa0e5597xx0/rtss_he ../alif/tests/drivers/dac/ \
+    -- -DDTC_OVERLAY_FILE=alif_dac0_max_current.overlay
+
+    # Minimum output current (0 uA) - light load testing
+    west build -b alif_e8_dk/ae822fa0e5597xx0/rtss_he ../alif/tests/drivers/dac/ \
+    -- -DDTC_OVERLAY_FILE=alif_dac0_min_current.overlay
```
Comment (VeijoPesonen) [inline tests/drivers/dac/README.rst]: We can assume building happens inside the sample dir to make documentation clearer. And the target board list is already on the yaml file so I don't see a reason to explicitly state how to make board selection. The User Guide already gives instructions how to do it. 
```suggestion
    # Unsigned mode (default): over-range rejection tests enabled
    west build -- -DDTC_OVERLAY_FILE=alif_dac0.overlay

    # Two's-complement mode: twos-complement tests enabled
    west build -- -DDTC_OVERLAY_FILE=alif_dac0_twoscomp.overlay

    # Maximum output current (1.5 mA) - HRM feature test
    west build -- -DDTC_OVERLAY_FILE=alif_dac0_max_current.overlay

    # Minimum output current (0 uA) - light load testing
    west build -- -DDTC_OVERLAY_FILE=alif_dac0_min_current.overlay
```

@Anilkumar-Alifsemi We should follow this style with all the sample and test rst files to avoid unnecessary repetition.

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
Small documentation adjustments to ease maintenance. Otherwise looks good to me.

@Anilkumar-Alifsemi FYI

File: tests/drivers/dac/README.rst [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+Requirement Coverage
+********************
+- **R-32_1**: Dual channel DAC control independently (test_dac_basic.c)
+- **R-32_2**: Unsigned binary / two's complement data (test_dac_data_format.c)
+- **R-32_3**: Continuous digital data conversion (test_dac_continuous.c)
+- **R-32_4**: Programmable load capacitance compensation (test_dac_capacitance.c)
+- **Boundary**: Edge case and boundary value tests (test_dac_boundary.c)
+
+Bug Regression Test Coverage
+****************************
```
Comment (Shamshoor-AlifSemi) [inline tests/drivers/dac/README.rst]: Bug regression test coverage is  not required in the README.rst

File: tests/drivers/dac/src/test_dac_basic.c [hunk 22]
```diff
@@ -0,0 +1,494 @@
...
+	 * or mask to 12 bits. Until the driver is fixed, skip these checks
+	 * to avoid locking in the buggy behavior as "passing".
+	 */
+	ret_1 = dac_write_value(dac_dev, DAC_CHANNEL_ID, 0x1FFF);
+	ret_2 = dac_write_value(dac_dev, DAC_CHANNEL_ID, 0xFFFF);
+
+	TC_PRINT("R-BUG-2 known issue: 0x1FFF returned %d, 0xFFFF returned %d "
+		 "(expected -EINVAL after driver fix)\n", ret_1, ret_2);
+
+	ztest_test_skip();
```
Comment (Shamshoor-AlifSemi) [inline tests/drivers/dac/src/test_dac_basic.c]: twoscomp_input_range_test writes values using dac_write_value then unconditionally skips the testcase at the end.

**Review by Shamshoor-AlifSemi — CHANGES_REQUESTED** [review]
Please address these comments.

**Review by VeijoPesonen — DISMISSED** [review]

**Review by VeijoPesonen — APPROVED** [review]

**Review by Shamshoor-AlifSemi — APPROVED** [review]

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
PR check failures

**Review by VeijoPesonen — APPROVED** [review]

File: tests/drivers/dac/README.rst (hunk 21 above)
Comment (VeijoPesonen) [reply tests/drivers/dac/README.rst]: Seem to be addressed.
